Smarty - When temporarily setting variable, fix cleanup for previously-undefined values.

totten opened the pull request:
Overview
----------------------------------------

You can temporarily assign Smarty values using various methods (`pushScope($vars)`/`popScope()` or `fetchWith(...$vars)` or `{crmScope ...}`). This expands the test-coverage for that mechanism.

Before
----------------------------------------

Fewer tests

After
----------------------------------------

Mo tests, mo problems.

Comments
----------------------------------------

Test expected to fail. Off-shoot from discussion with @demeritcowboy on #26432.

(Updated)
----------------------------------------

I've added a patch to fix the demonstrated failure.

totten committed:
(NFC) Smarty - Expand test coverage for scope-cleanup. Show inconsistency.

monishdeb commented:
@totten seems like test failure is related: 
```php
CRM_Core_SmartyTest::testFetchWith_CleanNonExistent
Failed asserting that true is false.

/home/homer/buildkit/build/build-3/web/sites/all/modules/civicrm/tests/phpunit/CRM/Core/SmartyTest.php:36
/home/homer/buildkit/build/build-3/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:238
/home/homer/buildkit/extern/phpunit8/phpunit8.phar:1721
```

totten committed:
CRM_Core_Smarty - Fix restoration of undefined values (following temporary assignments)

totten commented:
@monishdeb Yup, that was supposed to demonstrate the failure. I just pushed up the actual fix. Hopefully, fixing that doesn't cause massive trouble elsewhere...

demeritcowboy commented:
[CRM_Core_Smarty_plugins_CrmScopeTest.testBlank with data set #2](https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/2719/BKPROF=dfl,SUITES=phpunit-crm,label=bknix-tmp/testReport/junit/(root)/CRM_Core_Smarty_plugins_CrmScopeTest/testBlank_with_data_set__2/)
    [CRM_Core_Smarty_plugins_CrmScopeTest.testBlank with data set #4](https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/2719/BKPROF=dfl,SUITES=phpunit-crm,label=bknix-tmp/testReport/junit/(root)/CRM_Core_Smarty_plugins_CrmScopeTest/testBlank_with_data_set__4/)
    [CRM_Core_Smarty_plugins_CrmScopeTest.testBlank with data set #9](https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/2719/BKPROF=dfl,SUITES=phpunit-crm,label=bknix-tmp/testReport/junit/(root)/CRM_Core_Smarty_plugins_CrmScopeTest/testBlank_with_data_set__9/)

"Undefined array key 'x'"

I'm not clear why those haven't always failed. Possibly adding the useTransaction() stops the errors from getting swallowed somewhere. The fails are just a variation of all the warnings about smarty variables that never get assigned. So either, can't do those test cases, or the test would need updating to expect a php warning but do we really care to test that.

Also case 9 is the same as case 2. Probably some history there.

totten commented:
> I'm not clear why those haven't always failed. 

Right, I'm pretty sure it was a leak between test-cases. Consider case `#1` (`{crmScope x=1}{/crmScope}`):

* From global POV, you start with `x` undefined
* Then temporarily set `x=1`
* Then cleanup -- but the cleanup was flawed. It set `x=null` instead of unassigning `x`.

For all subsequent tests, they start with global `x=null` - so there's no notice about undefined `x`. The patch here changes the behavior -- subsequent tests now begin with `x` undefined. Hence, more warnings.

(One way to confirm the interaction -- checkout the older code, then play around with disabling or reordering specific `$cases`.)

(This suggests that `CRM_Core_Smarty` is currently a vector for other leaks... but that's a separate issue...)

>  expect a php warning but do we really care to enforce that.

Yeah, agree that doesn't need to be enforced.

We could define the global/default value. Since `x` uses a sequence (`x=1`, `x=2`, `x=3`) for each level of nesting, we could simply start with a global `x=0`. It doesn't provide any coverage for nullity, but 6996cbb56664012bf5ce0b5b04e8f3c7be63ff3f does.

> Also case 9 is the same as case 2. Probably some history there, or maybe it was testing that you could repeat a test and still get the same result after all the other pushing and popping.

Oh, I like that explanation... :)

totten committed:
CrmScopeTest - Use explicit value for default/global example

Before: We rely on default/global value of `x=UNDEFINED` or `x=NULL`.
But this is a bit tempermental.

After: Clear default.

Note: There's a prior commit which adds another test (CRM_Core_SmartyTest)
where the global value is specifically undefined or specifically null.
So that still has coverage.
